From 1224a89a07d9f3651dc2bdce4002b08792e7f006 Mon Sep 17 00:00:00 2001
From: PocketForge Worker <worker@pocketforge.invalid>
Date: Fri, 9 Oct 2026 19:05:19 -0400
Subject: [PATCH 13/50] hw/sd: add A100 SD host variant

Rebase resolutions:
- QOM: use the const class-init data signature required by v11 TypeInfo callbacks.

Assisted-by: LLM

diff --git a/hw/arm/Kconfig b/hw/arm/Kconfig
index 197b2e5c3c..01edf1eaa4 100644
--- a/hw/arm/Kconfig
+++ b/hw/arm/Kconfig
@@ -197,6 +197,7 @@ config POCKETFORGE_A133
     default y
     depends on TCG && AARCH64
     select ALLWINNER_I2C
+    select ALLWINNER_SDHOST
     select ARM_GIC
     select AXP2XX_PMU
     select DW_APB_UART
@@ -336,6 +337,7 @@ config ALLWINNER_H3
     select ALLWINNER_A10_PIT
     select ALLWINNER_SUN8I_EMAC
     select ALLWINNER_I2C
+    select ALLWINNER_SDHOST
     select ALLWINNER_WDT
     select SERIAL_MM
     select ARM_TIMER
diff --git a/hw/sd/Kconfig b/hw/sd/Kconfig
index 633b9afec9..d862cde442 100644
--- a/hw/sd/Kconfig
+++ b/hw/sd/Kconfig
@@ -10,6 +10,10 @@ config SSI_SD
 config SD
     bool
 
+config ALLWINNER_SDHOST
+    bool
+    select SD
+
 config SDHCI
     bool
     select SD
diff --git a/hw/sd/allwinner-sdhost.c b/hw/sd/allwinner-sdhost.c
index 87bb7df903..377ea94ce2 100644
--- a/hw/sd/allwinner-sdhost.c
+++ b/hw/sd/allwinner-sdhost.c
@@ -325,6 +325,14 @@ static void write_descriptor(AwSdHostState *s, hwaddr desc_addr,
                      MEMTXATTRS_UNSPECIFIED);
 }
 
+static hwaddr allwinner_sdhost_dma_addr(AwSdHostState *s, uint32_t addr)
+{
+    AwSdHostClass *sc = AW_SDHOST_GET_CLASS(s);
+
+    /* Reconstruct the byte address before discarding alignment bits. */
+    return ((hwaddr)addr << sc->dma_addr_shift) & (hwaddr)DESC_SIZE_MASK;
+}
+
 static uint32_t allwinner_sdhost_process_desc(AwSdHostState *s,
                                               hwaddr desc_addr,
                                               TransferDescriptor *desc,
@@ -361,7 +369,8 @@ static uint32_t allwinner_sdhost_process_desc(AwSdHostState *s,
         /* Write to SD bus */
         if (is_write) {
             dma_memory_read(&s->dma_as,
-                            (desc->addr & DESC_SIZE_MASK) + num_done, buf,
+                            allwinner_sdhost_dma_addr(s, desc->addr) + num_done,
+                            buf,
                             buf_bytes, MEMTXATTRS_UNSPECIFIED);
             sdbus_write_data(&s->sdbus, buf, buf_bytes);
 
@@ -369,7 +378,8 @@ static uint32_t allwinner_sdhost_process_desc(AwSdHostState *s,
         } else {
             sdbus_read_data(&s->sdbus, buf, buf_bytes);
             dma_memory_write(&s->dma_as,
-                             (desc->addr & DESC_SIZE_MASK) + num_done, buf,
+                             allwinner_sdhost_dma_addr(s, desc->addr) + num_done,
+                             buf,
                              buf_bytes, MEMTXATTRS_UNSPECIFIED);
         }
         num_done += buf_bytes;
@@ -385,7 +395,7 @@ static uint32_t allwinner_sdhost_process_desc(AwSdHostState *s,
 static void allwinner_sdhost_dma(AwSdHostState *s)
 {
     TransferDescriptor desc;
-    hwaddr desc_addr = s->desc_base;
+    hwaddr desc_addr = allwinner_sdhost_dma_addr(s, s->desc_base);
     bool is_write = (s->command & SD_CMDR_WRITE);
     uint32_t bytes_done = 0;
 
@@ -418,7 +428,7 @@ static void allwinner_sdhost_dma(AwSdHostState *s)
         if (desc.status & DESC_STATUS_LAST) {
             break;
         } else {
-            desc_addr = desc.next;
+            desc_addr = allwinner_sdhost_dma_addr(s, desc.next);
         }
     }
 
@@ -939,6 +949,18 @@ static void allwinner_sdhost_sun50i_a64_emmc_class_init(ObjectClass *klass,
     sc->can_calibrate = true;
 }
 
+static void allwinner_sdhost_sun50i_a100_class_init(ObjectClass *klass,
+                                                    const void *data)
+{
+    AwSdHostClass *sc = AW_SDHOST_CLASS(klass);
+
+    /* Matches Linux's allwinner,sun50i-a100-mmc (sun20i_d1_cfg). */
+    sc->max_desc_size = 8 * KiB;
+    sc->dma_addr_shift = 2;
+    sc->is_sun4i = false;
+    sc->can_calibrate = true;
+}
+
 static const TypeInfo allwinner_sdhost_info = {
     .name          = TYPE_AW_SDHOST,
     .parent        = TYPE_SYS_BUS_DEVICE,
@@ -973,6 +995,12 @@ static const TypeInfo allwinner_sdhost_sun50i_a64_emmc_info = {
     .class_init    = allwinner_sdhost_sun50i_a64_emmc_class_init,
 };
 
+static const TypeInfo allwinner_sdhost_sun50i_a100_info = {
+    .name          = TYPE_AW_SDHOST_SUN50I_A100,
+    .parent        = TYPE_AW_SDHOST,
+    .class_init    = allwinner_sdhost_sun50i_a100_class_init,
+};
+
 static const TypeInfo allwinner_sdhost_bus_info = {
     .name = TYPE_AW_SDHOST_BUS,
     .parent = TYPE_SD_BUS,
@@ -987,6 +1015,7 @@ static void allwinner_sdhost_register_types(void)
     type_register_static(&allwinner_sdhost_sun5i_info);
     type_register_static(&allwinner_sdhost_sun50i_a64_info);
     type_register_static(&allwinner_sdhost_sun50i_a64_emmc_info);
+    type_register_static(&allwinner_sdhost_sun50i_a100_info);
     type_register_static(&allwinner_sdhost_bus_info);
 }
 
diff --git a/hw/sd/meson.build b/hw/sd/meson.build
index b43d45bc56..7c644ca40f 100644
--- a/hw/sd/meson.build
+++ b/hw/sd/meson.build
@@ -7,6 +7,6 @@ system_ss.add(when: 'CONFIG_SSI_SD', if_true: files('ssi-sd.c'))
 system_ss.add(when: 'CONFIG_OMAP', if_true: files('omap_mmc.c'))
 system_ss.add(when: 'CONFIG_RASPI', if_true: files('bcm2835_sdhost.c'))
 system_ss.add(when: 'CONFIG_ASPEED_SOC', if_true: files('aspeed_sdhci.c'))
-system_ss.add(when: 'CONFIG_ALLWINNER_H3', if_true: files('allwinner-sdhost.c'))
+system_ss.add(when: 'CONFIG_ALLWINNER_SDHOST', if_true: files('allwinner-sdhost.c'))
 system_ss.add(when: 'CONFIG_NPCM7XX', if_true: files('npcm7xx_sdhci.c'))
 system_ss.add(when: 'CONFIG_CADENCE_SDHCI', if_true: files('cadence_sdhci.c'))
diff --git a/include/hw/sd/allwinner-sdhost.h b/include/hw/sd/allwinner-sdhost.h
index d154d81d24..40a258c393 100644
--- a/include/hw/sd/allwinner-sdhost.h
+++ b/include/hw/sd/allwinner-sdhost.h
@@ -44,6 +44,9 @@
 /** Allwinner sun50i-a64 emmc */
 #define TYPE_AW_SDHOST_SUN50I_A64_EMMC  TYPE_AW_SDHOST "-sun50i-a64-emmc"
 
+/** Allwinner sun50i-a100 SD/MMC */
+#define TYPE_AW_SDHOST_SUN50I_A100 TYPE_AW_SDHOST "-sun50i-a100"
+
 /** @} */
 
 /**
@@ -137,6 +140,7 @@ struct AwSdHostClass {
 
     /** Maximum buffer size in bytes per DMA descriptor */
     size_t max_desc_size;
+    uint8_t dma_addr_shift;
     bool   is_sun4i;
 
     /** does the IP block support autocalibration? */
-- 
2.43.0

